pkg/api/grpc_server.go

```diff
@@ -0,0 +1,204 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package api
+
+import (
+	"context"
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/coreos/go-oidc/v3/oidc"
+	empty "github.com/golang/protobuf/ptypes/empty"
+	certauth "github.com/sigstore/fulcio/pkg/ca"
+	"github.com/sigstore/fulcio/pkg/challenges"
+	"github.com/sigstore/fulcio/pkg/config"
+	"github.com/sigstore/fulcio/pkg/ctl"
+	fulciogrpc "github.com/sigstore/fulcio/pkg/generated/protobuf"
+	"github.com/sigstore/fulcio/pkg/log"
+	"github.com/sigstore/sigstore/pkg/cryptoutils"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/metadata"
+)
+
+type grpcCAServer struct {
+	fulciogrpc.UnimplementedCAServer
+	ct ctl.Client
+	ca certauth.CertificateAuthority
+}
+
+func NewGRPCCAServer(ct ctl.Client, ca certauth.CertificateAuthority) fulciogrpc.CAServer {
+	return &grpcCAServer{
+		ct: ct,
+		ca: ca,
+	}
+}
+
+const (
+	MetadataOIDCTokenKey = "oidcidentitytoken"
+)
+
+func (g *grpcCAServer) CreateSigningCertificate(ctx context.Context, request *fulciogrpc.CreateSigningCertificateRequest) (*fulciogrpc.SigningCertificate, error) {
+	logger := log.ContextLogger(ctx)
+
+	// OIDC token either is passed in gRPC field or was extracted from HTTP headers
+	token := ""
+	if request.Credentials != nil {
+		token = request.Credentials.GetOidcIdentityToken()
+	}
+	if token == "" {
+		if md, ok := metadata.FromIncomingContext(ctx); ok {
+			vals := md.Get(MetadataOIDCTokenKey)
+			if len(vals) == 1 {
+				token = vals[0]
+			}
+		}
+	}
+
+	principal, err := authorize(ctx, token)
+	if err != nil {
+		return nil, handleFulcioGRPCError(ctx, codes.Unauthenticated, err, invalidCredentials)
+	}
+
+	publicKeyBytes := request.PublicKey.Content
+	// try to unmarshal as DER
+	publicKey, err := x509.ParsePKIXPublicKey([]byte(publicKeyBytes))
+	if err != nil {
+		// try to unmarshal as PEM
+		logger.Debugf("error parsing public key as DER, trying pem: %v", err.Error())
+		publicKey, err = cryptoutils.UnmarshalPEMToPublicKey([]byte(publicKeyBytes))
+		if err != nil {
+			return nil, handleFulcioGRPCError(ctx, codes.InvalidArgument, err, invalidPublicKey)
+		}
+	}
+
+	// Validate public key, checking for weak key parameters.
+	if err := cryptoutils.ValidatePubKey(publicKey); err != nil {
+		return nil, handleFulcioGRPCError(ctx, codes.InvalidArgument, err, insecurePublicKey)
+	}
+
+	subject, err := challenges.ExtractSubject(ctx, principal, publicKey, request.ProofOfPossession)
+	if err != nil {
+		return nil, handleFulcioGRPCError(ctx, codes.InvalidArgument, err, invalidSignature)
+	}
+
+	var csc *certauth.CodeSigningCertificate
```

Comment:
Merge conflict incoming: This section will have to be updated based on https://github.com/sigstore/fulcio/pull/507/files#diff-41ae48039acc4fc3fe73112a8e3af14bad4ab8691f9edec07804e0fd9604bec4

Reply:
he who merges first wins!

Reply:
I think this one will go in first right? #507 needs to wait until Cosign 1.8

Reply:
We can submit #507 now, the intermediate CA won't be rolled out until we make the changes to the production configs.